aip/general/0193.md

```diff
@@ -75,132 +86,191 @@ following:
-  in one location, `LocalizedMessaage.message`, across all methods of
-  the API Service.
-
-Reasons to present an error message in `LocalizedMessage.message` that
-differs from `Status.message` include the following:
+##### Uniqueness
 
-- The service requires an end-user facing error message that differs
-  from the "debug message".
-- Ongoing, iterative error message improvements are expected.
+Each type of detail payload **must** be included at most once. For
+example, there **must not** be more than one [`BadRequest`][BadRequest]
+message in the `details`, but there **may** be a `BadRequest` and a
+[`PreconditionFailure`][PreconditionFailure].
 
-When including `LocalizedMessage`, both fields, `locale` and `message`,
-**must** be populated.  If the service is to be localized, the value of
-`locale` **must** change dynamically. See,
-"[Localization](#localization)". Otherwise, `locale` **must** always
-present the service's default locale, e.g. "en-US".
 
-When adding an error message via `LocalizedMessage`, `ErrorInfo`
-**must** be introduced either before or at the same time. If there are
-dynamic segments found in the text, the values of these variables
-**must** be included in `ErrorInfo.metadata`.  See, "[Dynamic
-variables](#dynamic-variables)".
+##### ErrorInfo payload
 
-**Warning:** If `LocalizedMessage` is released without `ErrorInfo`, the
-same [concerns](#status-message-warning) regarding client misuse of
-textual error messages apply.
+The [`ErrorInfo`][ErrorInfo] message is the required way to send a
+machine-readable identifier. All error responses **must** include an
+`ErrorInfo` payload in `details`.  Variable information **should** be
+included in `metadata` in `ErrorInfo` and **must** be included if it
+appears within an error message.
 
-#### Dynamic variables
+When introducing an error that represents a failure scenario that did
+not previously occur for the service, the payload **must** include
+`ErrorInfo` and any variables found in dynamic segments of the error
+message **must** be present in `ErrorInfo.metadata` (See [Dynamic
+variables](#dynamic-variables).
+
+**Note:** `ErrorInfo` represents a special case. There **must** be
+exactly one `ErrorInfo`. It is required.
+
+[ErrorInfo Payload][ErrorInfo]
+
+Details of this object are summarized in the following fields, field
+descriptions, and examples:
+
+- `reason string`
+  - A short snake_case description of why the error occurred. Error reasons
+    are unique within a particular domain of errors. The error reason
+    **must** do the following:
+  - Be at most 63 characters and match a regular expression of
+    \`\[A-Z\]\[A-Z0-9\_\]+\[A-Z0-9\]\`, which represents
+    UPPER_SNAKE_CASE.
+  - Be unique and consumable by machine actors for automation.
+  - Be meaningful enough for a human reader to understand what the
+    reason refers to.
+  - Map to one and only one cause, That is, avoid making machine actors
+    reliant on ErrorInfo.
```

Comment:
I don't know what this means. Please clarify.

Reply:
I think this is trying to say don't put dynamic data in here that can't be read from `error_info.metadata`.

@shwoodard should this say that any dynamic values must be present in `error_info.metadata`? It feels a bit odd to define it that way but [LocalizedMessage](https://github.com/googleapis/googleapis/blob/13c6878a4c2822c6db7b986e717bf57943bf1681/google/rpc/error_details.proto#L277) doesn't have its own metadata value.

I think the bigger question here is, what's the purpose of `LocalizedMessage` relative to `ErrorInfo`? Specifically, what should `error_info.message` be used for, and what should `localized_message.message` be used for? The latter I think is what we want to display to users, but is there any use for the former, except for legacy applications?

Reply:
From the text "The LocalizedMessage payload should contain the complete resolution to the error."
Shouldn't this be a "must?"

Reply:
Please remove this statement as it's confusing and does not add anything, this above statement is actually sufficient:

>   + Be unique and consumable by machine actors for automation.

Reply:
Done.

Reply:
Sorry for the confusion, what I meant was that this statement is the one that's sufficient and we should keep:

> Be unique and consumable by machine actors for automation.

This statement, that's highlighted by this comment, is the one we should remove:

>  - Map to one and only one cause, That is, avoid making machine actors
    reliant on ErrorInfo.